### PR TITLE
Loadout strip

### PR DESCRIPTION
**Pull request by nev-r**
looking for some early feedback on this.
needs to incorporate pieces of #9306,
needs application preparation function made,
needs a lot of CSS. not attached to the icons or the outline, or much of anything really.

<hr/>

stops treating in-game loadouts as first-class loadouts on the loadouts page.
replaces them instead with 10 slots, and ways to manage them and check them

image

identifies whether a slot is
 - currently equipped
 - "is saved" already as a dim loadout (matches all the requirements of a DIM loadout)
 - is equippable (likely to succeed if the user attempts to equip it ingame)
 - 
image

 
offers more options in both a quick dropdown
image

and a full sheet
image

### CODE REVIEW

File: src/app/loadout/ingame/ingame-loadout-utils.ts [hunk 1]
```diff
@@ -36,3 +39,68 @@ export function useItemsFromInGameLoadout(loadout: InGameLoadout) {
...
+      if (!i.equip) {
+        return false;
+      }
+      const itemType = defs.InventoryItem.get(i.hash).itemType;
+      // only checking the items that game loadouts support
+      return (
+        itemType === DestinyItemType.Weapon ||
+        itemType === DestinyItemType.Armor ||
+        itemType === DestinyItemType.Subclass
+      );
```
Comment (robojumper) [inline src/app/loadout/ingame/ingame-loadout-utils.ts:65]: This looks a bit like `gameLoadoutCompatibleBuckets` above.

File: src/app/loadout/ingame/ingame-loadout-utils.ts [hunk 2]
```diff
@@ -36,3 +39,68 @@ export function useItemsFromInGameLoadout(loadout: InGameLoadout) {
...
+ * does this game loadout, meet the requirements of this DIM loadout:
+ *
+ * does it include the items the DIM loadout would equip,
+ * and represent a full application of the DIM loadout's required mods?
+ */
+export function implementsDimLoadout(
+  inGameLoadout: InGameLoadout,
+  dimLoadout: Loadout,
+  defs: D2ManifestDefinitions
+) {
```
Comment (robojumper) [inline src/app/loadout/ingame/ingame-loadout-utils.ts:53]: This function should probably be called with a fully resolved loadout (including resolved loadout mods) 

1. DIM loadouts have item migration, e.g. for crafted weapons, where the IDs don't match and instead the crafted timestamp is used.
2. We just merged some loadout mod migration in #9287, so some old loadouts now have old stat mod hashes but resolve to new ones. `getModsFromLoadout` takes care of that.
3. DIM loadouts always store the regular version of a mod and apply what's available, while an in-game loadout may store the reduced-cost copy. This function should probably involve https://github.com/DestinyItemManager/DIM/blob/73f2f5e705491f09b6d3910f830bb4184047e931/src/app/loadout/mod-utils.ts#L128-L133 when checking mods.

Comment (robojumper) [reply src/app/loadout/ingame/ingame-loadout-utils.ts:53]: To clarify, for the items you probably don't need a ResolvedLoadoutItem, just need to call `findItemForLoadout` to find the correct item.

**Comment by bhollis** [comment]
I like it!

What do you think about not showing the empty loadout slots?

**Comment by nev-r** [comment]
pretty ambivalent! just wanted something to visually encourage people to use in-game loadouts, but there could be some other kind of button or message here, or nothing at all. can't make a horse drink in-game loadouts

File: src/app/loadout/ingame/ingame-loadout-utils.ts (hunk 2 above)
Comment (nev-r) [reply src/app/loadout/ingame/ingame-loadout-utils.ts]: ok this now uses resolved mods and items. hopefully game-side loadout plays nice with mod discounts too. stuff is accurately detecting though, in testing here.

Comment (nev-r) [reply src/app/loadout/ingame/ingame-loadout-utils.ts]: still need to screw around with activating/deactivating artifact discounts but so far matching remarkably well using the mod resolver.

**Comment by nev-r** [comment]
same objections re overall page layout. filters area really needs a label imo. 

re: names
thing is, loadout names suck.
- they're very arbitrary and the user has little control over them
- picking a name at all is a major compromise with bungie, and often not indicative of much, if anything
- they are _very_ secondary identifiers, you can't even see them ingame without hovering.
skillful/proper use of in-game loadouts _necessarily_ relies on remembering positioning, number, and icon.

really don't believe it's useful to promote them to top-level identifiers in DIM, when there is no corresponding information in D2.

**Comment by liminalAce** [comment]
Fair enough. It's all arbitrary really was more so just giving it another level for ID. But I take your point that the game doesn't.

**Comment by nev-r** [comment]
think i'd rather lose the dropdown menu, than make most of the whole rectangle inert to clicks, that seems no fun.

that said if every action relies on bringing up a busy details sheet, i can imagine wanting, or people asking for, faster shortcuts to doing those actions. kinda stuck in the middle here. thoughts on what we can do to justify two different modes of operation and give them reasonable entry points?

could be a `Details` in the dropdown, that also clicking the main bulk leads to.. maybe that doesn't address your problem with it.

**Comment by liminalAce** [comment]
I think the whole rectangle should be clickable. Basically treat it like the emblems. The only reason the dots are there at all is because people didn't know it was a context menu. So basically I'm saying: 

Remove the big click that opens the sheet, and instead have the big click and the dots click open the dots menu but have an option in the dots menu for opening the sheet. 

So one big click area, does all the things.

File: src/app/dim-ui/FilterPills.tsx [hunk 3]
```diff
@@ -53,6 +56,7 @@ export default function FilterPills({
       className={clsx(styles.guide, className, { [styles.darkBackground]: darkBackground })}
       onClick={clearSelection}
     >
+      {includeLabel && <h3>{t('Loadouts.Filters')}</h3>}
```
Comment (bhollis) [inline src/app/dim-ui/FilterPills.tsx:59]: This doesn't belong here - filter pills are generic and not just for loadouts

File: src/app/inventory/item-move-service.ts [hunk 4]
```diff
@@ -1075,32 +1082,53 @@ export function sortMoveAsideCandidatesForStore(
...
-      compareBy((i) => item && i.bucket.hash === item.bucket.hash),
-      // or at least same category (weapons, armor)
-      compareBy((i) => item && i.bucket.sort === item.bucket.sort),
-      // Always prefer keeping something that was manually moved where it is
-      compareBy((i) => -getLastManuallyMoved(i)),
+      compareBy((displaced) => !displaced.equipped),
+      // prefer same bucket over everything, because that makes space in the correct "pocket"
+      compareBy((displaced) => displacer && displaced.bucket.hash === displacer.bucket.hash),
+
+      // THIS SEEMS LIKE SOMETHING FROM D1
```
Comment (bhollis) [inline src/app/inventory/item-move-service.ts:1102]: Indeed - in D1 engrams exist within normal inventory.

File: src/app/inventory/item-move-service.ts [hunk 5]
```diff
@@ -1075,32 +1082,53 @@ export function sortMoveAsideCandidatesForStore(
...
-      compareBy((i) => !i.equipped),
-      // prefer same bucket over everything
-      compareBy((i) => item && i.bucket.hash === item.bucket.hash),
-      // or at least same category (weapons, armor)
-      compareBy((i) => item && i.bucket.sort === item.bucket.sort),
-      // Always prefer keeping something that was manually moved where it is
-      compareBy((i) => -getLastManuallyMoved(i)),
+      compareBy((displaced) => !displaced.equipped),
+      // prefer same bucket over everything, because that makes space in the correct "pocket"
+      compareBy((displaced) => displacer && displaced.bucket.hash === displacer.bucket.hash),
```
Comment (bhollis) [inline src/app/inventory/item-move-service.ts:1100]: I suppose this doesn't need to be a constraint if the `fromStore` is the vault.

File: src/app/loadout/ingame/selectors.ts [hunk 6]
```diff
@@ -28,3 +105,52 @@ export const inGameLoadoutsForCharacterSelector = createSelector(
...
+  ) => {
+    const selectedStore = getStore(stores, storeId)!;
+    return (
+      inGameLoadouts
+        // seems unlikely the game would return valid, itemful loadouts for slot you havne't earned, but we respect this setting.
+        // inGameLoadoutsForCharacterSelector filters out empty loadouts, so we have to go by their self-stated index, not array length
+        .filter((gameLoadout) => gameLoadout.index < availableLoadoutSlots)
+        .map((gameLoadout) => {
+          const isEquippable = gameLoadout.items.every((li) => {
+            const liveItem = allItems.find((di) => di.id === li.itemInstanceId);
```
Comment (bhollis) [inline src/app/loadout/ingame/selectors.ts:133]: Consider using `potentialLoadoutItemsByItemId`

File: src/app/inventory/item-move-service.ts (hunk 5 above)
Comment (nev-r) [reply src/app/inventory/item-move-service.ts:1100]: agreed and that would be a big improvement for once this algo also takes into account amount of free space

**Comment by bhollis** [comment]
Hmm, this last change has a bit of a bug - the converter I made for sharing/copy-to-DIM-loadout strips out empty plugs (since we don't have a great way to represent those in DIM loadouts yet) but I probably don't want to do that when *displaying* in-game loadouts since it really will reset those.

**Comment by ryan-rushton** [comment]
Something is off with the currently equipped icon

Screenshot 2023-04-10 at 5 50 15 pm

**Comment by ryan-rushton** [comment]
Also appears like you can't equip them at this stage? Not sure if a bug or a TODO

File: src/app/loadout-drawer/loadouts-selector.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { currentProfileSelector } from 'app/dim-api/selectors';
+import { RootState } from 'app/store/types';
+import { emptyArray } from 'app/utils/empty';
+import { createSelector } from 'reselect';
+import { convertDimApiLoadoutToLoadout } from './loadout-type-converters';
+import { Loadout } from './loadout-types';
+
+// had to pull this out to another file because things got weird :(
```
Comment (ryan-rushton) [inline src/app/loadout-drawer/loadouts-selector.ts:8]: Cyclic imports?

**Pull request merged**